슬라이드 뷰어 MVP구현

aereeeee opened the pull request:
# 슬라이드 뷰어 MVP구현

## 해당 이슈 📎

- 슬라이드를 앞뒤로 넘길 수 있다 (#42)

## 변경 사항 🛠
#### 구현 내용 요약
- 클라이언트에서 server-converter에 pdf업로드 시 응답으로 slideUrls 배열을 줍니다.
- slideUrls를 server-api에서 DB에 저장합니다.
- 클라이언트는 쿼리한 채널정보를 context API를 통해 채널 페이지에 제공합니다.
- 슬라이더 뷰어의 인디케이터 클릭시 페이지 변환이 일어납니다.
#### 그 외 참고사항
- 공통으로 쓸수 있는 useChannelSelector 훅을 만들었습니다.
- context를 컴포넌트에서 직접적으로 import하지 않고 훅을 통해 원하는 state를 가져 올 수 있습니다.

## 테스트 ✨

> 없음

## 리뷰어 참고 사항 🙋‍♀️

- server-api, web에서 channel관련 파일만 보시면 됩니다.
- server-converter에서 upload미들웨어와 app.ts부분만 보시면 됩니다.

## 앞으로 해야할일..
- 현재 최소단위로 구현했기 때문에 디테일한 UI를 마저 구현할 것입니다.(슬라이드 겹쳐진 UI , 페이지 변환 애니메이션)
- 스피커와 리스너의 슬라이드 동기화를 구현하고 있습니다🔥

aereeeee committed:
[ STYLE ] 채널 페이지 UI 구현을 위해 컴포넌트 초기 레이아웃을 설계 (#42)

Why
- 구체적인 컴포넌트 개발에 들어가기 전, 전체 페이지의 기본적인 레이아웃을 잡을 필요가 있다.
- 특히 width, height가 브라우저에 의존하므로 컴포넌트의 크기가 기본적으로 비율로 결정되어야한다.

What
- Channel페이지를 구성하는 ToolBar, Slide, Chat컴포넌트를 나누었다.
- flex속성을 이용하여 레이아웃을 설계하였다.

aereeeee committed:
[ FEATURE ] 슬라라이더 뷰어 Indicator UI 구현 (#42)

Why
- 슬라이더의 왼쪽과 오른쪽에 마우스를 오버했을 때 슬라이더 이동 indicator가 나와야 한다.
- 평상 시에는 Indicator의 방해 없이 슬라이드 화면을 볼 수 있어야 한다.

What
- Indicator 컴포넌트를 만들었다.
- direction을 props로 전달하여 각각 back, foward인디케이터에 마우스 hover시 나타나도록 구현하였다.

aereeeee committed:
[ FEATURE ] 채널 접속 시 슬라이드이미지와 파일의 url을 쿼리하도록 관련 로직 수정 (#42)

Why
- 채널 접속 시 슬라이드 이미지를 보여주어야 한다.
- 오브젝트 스토리지의 슬라이드 이미지 url 데이터가 필요하다.

What
- server-converter에서 파일과 슬라이드를 업로드 하고 location을 reslove하도록 구현
- 미들웨어에서 request에 파일과 슬라이드를 등록하고 모든 미들웨어 처리 후 응답하도록 구현
- 클라이언트 파일 업로드 시 server-converter에서 응답한 url을 mutate에 담아 server-api에서 처리하도록 관련 graphQL로직 수정
- server-converter에서 파일 업로드 시 전체 공개 권한을 갖게 하기 위해 ACL 파라미터 추가

aereeeee committed:
[ FEATURE ] Channel페이지에서 slideUrls 데이터도 provide하도록 구현 (#42)

Why
- 채널페이지 하위 컴포넌트들은 슬라이드 이미지 url데이터를 쉽게 알수있어야 한다.
- 여러 컴포넌트에서 context에 직접 접근하지 않고 컨텍스트 데이터를 사용할 수 있어야 한다.

What
- ChannelContext.Provider에서 slideUrls를 내려주도록 구현
- useChannelSelector 훅을 만들어 컴포넌트에서 해당 훅을 이용하여 컨텍스트 데이터에 접근하도록 구현

aereeeee committed:
[ FEATURE ] Slide indicator 클릭 시 페이지 변환 구현 (#42)

Why
- 인디케이터를 통해 페이지가 앞 뒤로 넘어가야한다,
- 첫 장과 마지막 장일 경우 이전, 다음 페이지로 이동하지 않아야 한다.

What
- slideView컴포넌트에서 useState로 페이지 state를 관리하도록 구현
- 인디케이터 클릭시 handleSetPage 핸들러로 페이지 변환 구현

aereeeee committed:
[ STYLE ] 슬라이드 이미지 사이즈에 맞게 슬라이브 뷰어 사이즈 변경 (#42)

Why
- 스피커가 제공한 슬라이드의 비율에 맞게 슬라이더뷰어가 제공되어야 한다.

What
- MainSlide컴포넌트에서 이미지 사이즈를 auto와 max-width, max-height 100%로 조절하였다
- 컴포넌트 역할을 명확히 하기 위해 SlideView컴포넌트를 SlideViewer컴포넌트로 이름을 변경하였다

always-awake commented on server-converter/src/middlewares/upload.ts:
```diff
@@ -14,28 +14,43 @@ const uploadToObjectStorage = (
 const uploadMiddleware: RequestHandler = (req, _, next) => {
   const { channelId } = req.body;
-  const uploadFile = uploadToObjectStorage(req.file.path, channelId, 'index', true);
-  const uploadSlides = req.slides.map((slide) => uploadToObjectStorage(slide.path, channelId, `${slide.page}`, false));
```
적절하게 적용해준 줄바꿈 보기 편하네요🥰

devjaewon commented on server-api/src/apollo/resolvers/channels.js:
```diff
@@ -2,26 +2,53 @@ const { ApolloError } = require('apollo-server-express');
+  slideUrls,
+  fileUrl,
 });
 
```
회의 이전에 작업한 내용이라 아직 반영이 안된 부분인 듯 싶습니다.
추후 리팩토링 시 `channel.toPayload()` 메서드로 바로 변환해서 넣으면 좋을 것 같습니다 😊

aereeeee replied:
넵~

always-awake commented on web/src/hooks/useCheckChannel.js:
```diff
@@ -6,10 +6,13 @@ const CHECK_CHANNEL = gql`
   } 
 `;
 const useCheckChannel = (channelId) => {
+  console.log('checkchannel');
```
console!!!! 지워주세용 😭

aereeeee replied:
넵!

happydhKim commented on web/src/components/channel/Slide/SlideViewer/SlideViewer.jsx:
```diff
@@ -0,0 +1,35 @@
+    if (direction === 'back' && page === 0) return;
+    if (direction === 'foward' && page === slideUrls.length - 1) return;
+    if (direction === 'back') setPage(page - 1);
+    else setPage(page + 1);
```
if문이 많아서 코드가 안예뻐지는 거 같은데 어떻게 생각하시나요??🤔
크롱마스터님께서 말씀하신 것처럼 함수 하나로 빼서 하면 훨씬 예뻐질거 같아요. 아래 제가 예시로 한거보다 좋은게 있지 않을까요??🐤
```javascript
const movePossible = (direction, page) => (
  (direction === 'back' && page === 0) 
  || (direction === 'foward' && page === slideUrls.length - 1))
  ? false : true;
```

aereeeee replied:
와우 좋습니다 감사합니다~

happydhKim commented on web/src/hooks/useCheckChannel.js:
```diff
@@ -6,10 +6,13 @@ const CHECK_CHANNEL = gql`
   } 
 `;
 const useCheckChannel = (channelId) => {
+  console.log('checkchannel');
```
테스트 용도라면 PR을 날릴 때는 지워주는게 좋지 않을까요?🐤

aereeeee replied:
넵 ㅜㅜ

happydhKim reviewed (approved):
전반적으로 깔끔하고 팀을 위해서 노력하시는게 너무 좋아요!
스타일쪽 px관련은 우리 팀회의를 통해서 빨리 논의 해봐요!!🐶

always-awake commented on web/src/components/channel/Slide/style.jsx:
```diff
@@ -0,0 +1,11 @@
+import styled from 'styled-components';
+
+export default {
+  SlideWrapper: styled.div`
```
스타일 컴포넌트 타입체크해주세용 😱

aereeeee replied:
특정 attribute를 지정해야할때 필요한 내용이라 패스하겠습니다~

always-awake commented on web/src/pages/Channel/Channel.jsx:
```diff
@@ -21,6 +21,8 @@ const Channel = () => {
       value={{ isMaster: data.isMaster }}
     >
       <S.ChannelWrapper>
```
팀의 컨벤션에 따라 S.Wrapper 로 통일해주세요옹

aereeeee replied:
이 컨벤션은 재원님 의견에 따라 컴포넌트 이름으로 가겠습니다~

always-awake commented on web/src/pages/Channel/style.jsx:
```diff
@@ -2,7 +2,8 @@ import styled from 'styled-components';
-    flex: 1;
+    background-color:  ${({ theme }) => theme.palette.background.main};
+    flex: 1 1 auto;
+    display:flex;
```
띄어쓰기!

aereeeee replied:
넵!

always-awake reviewed (approved):
간단한 띄어쓰기(스타일 컴포넌트 내), 스타일 컴포넌트 타입체크 정도만 해주시면 좋을것 같습니다!! 
코드 줄바꿈, 읽기 쉬운 변수 명등으로 읽기 편했던 코드였던 것 같습니다! 수고 많으셨습니다 ㅎㅎ 😘

aereeeee committed:
[ REFACTOR ] 이슈 #42에 대한 PR 리뷰사항 반영

Why
- 코드 컨벤션을 맞출 필요가 있다.
- 함수를 더 작게 나누는 것이 좋다.

What
- Wrapper컴포넌트를 팀 컨벤션에 맞게 해당 컴포넌트 이름으로 변경
- movePagePossible유틸로 현재 페이지가 이동가능한지 체크하는 로직을 분리

aereeeee committed:
Merge branch 'develop' into feature/issue42
